### DISCUSSION TITLE
Why injected EntityManager performs insert instead of update?

### DISCUSSION THREAD

**Discussion by MartkCz** [spiral/spiral#1152] [Q&A] [answered]
I normally use DI for everything, but I kept getting exceptions with duplicate keys. is this really expected behavior?

```php
class Model
{

	public function __construct(
		private ORM $orm,
		private EntityManagerInterface $em,
	)
	{
	}

	public function method(Symbol $symbol): void
	{
		/** @var ShortTimeSeriesMetadata|null $metadata */
		$metadata = $this->orm->getRepository(ShortTimeSeriesMetadata::class)
			->findByPK($symbol->value);

		if (!$metadata) {
			return;
		}

		$metadata->setLastTry(new DateTimeImmutable());

		// performs insert operation
		$this->em->persist($metadata);
		$this->em->run();
		
		// performs update operation		
		$em = new EntityManager($this->orm);
		$em->persist($metadata);
		$em->run();
	}

}
```

**Comment by roxblnfk** [comment, marked as answer]
Looks like you resolve ORM via class, not interface. That's why you get a new ORM with a clean heap ant it is trying to insert the entity because it looks new for the ORM

**Reply by MartkCz** [reply]
I made a mistake in the example code. Real code was using ORMInterface

**Reply by roxblnfk** [reply]
Do you have any ideas why a different instance of ORM is getting into the EntityManager?

**Reply by MartkCz** [reply]
The bug was outside the framework, but your answer helped a lot.
